react/src/page_template/settings.ts

```diff
@@ -96,35 +115,151 @@ export class Settings {
         useEffect(() => {
             setResult(this.getMultiple(keys)) // So that if `key` changes we change our result immediately
             const observer = (): void => { setResult(this.getMultiple(keys)) }
-            keys.forEach(key => this.observers.get(key).add(observer))
+            keys.forEach(key => this.settingValueObservers.get(key).add(observer))
             return () => {
-                keys.forEach(key => this.observers.get(key).delete(observer))
+                keys.forEach(key => this.settingValueObservers.get(key).delete(observer))
             }
         // eslint-disable-next-line react-hooks/exhaustive-deps -- Our dependencies are the keys
         }, keys)
         return result
     }
 
     setSetting<K extends keyof SettingsDictionary>(key: K, newValue: SettingsDictionary[K]): void {
-        this.settings[key] = newValue
-        localStorage.setItem('settings', JSON.stringify(this.settings))
-        this.observers.get(key).forEach((observer) => { observer() })
+        if (this.stagedSettings !== undefined && (key in this.stagedSettings)) {
+            this.stagedSettings[key] = newValue
+        }
+        else {
+            this.settings[key] = newValue
+            localStorage.setItem('settings', JSON.stringify(this.settings))
+        }
+        this.settingValueObservers.get(key).forEach((observer) => { observer() })
     }
 
     updateSetting<K extends keyof SettingsDictionary>(key: K, makeNewValue: (oldValue: SettingsDictionary[K]) => SettingsDictionary[K]): void {
         this.setSetting(key, makeNewValue(this.get(key)))
     }
 
     get<K extends keyof SettingsDictionary>(key: K): SettingsDictionary[K] {
+        if (this.stagedSettings !== undefined && (key in this.stagedSettings)) {
+            return this.stagedSettings[key]!
+        }
         return this.settings[key]
     }
 
     getMultiple<K extends keyof SettingsDictionary>(keys: K[]): Pick<SettingsDictionary, K> {
-        return Object.fromEntries(keys.map(key => [key, this.settings[key]])) as Pick<SettingsDictionary, K>
+        return Object.fromEntries(keys.map(key => [key, this.get(key)])) as Pick<SettingsDictionary, K>
     }
 
     // Singular settings means we can use observers
     static Context = createContext(new Settings())
+
+    /**
+     * Staged Mode
+     */
+    private stagedSettings?: Partial<SettingsDictionary>
+
+    /**
```

Comment:
Comment should explain what staging is

Reply:
Added an explanation, lmk if you have a better one

Reply:
LG